File: AOC2021.cpp/Day1.cpp
#include "Day1.h"
#include <fstream>
#include <sstream>
#include <vector>
#include <iostream>
#include <numeric>

int Day1_1(std::vector<int> _dayData);
int Day1_2(std::vector<int> _dayData);

int Day1::Resolve(int _part, std::string _filename)
{
	std::vector<int> data;
	{
		std::ifstream file(_filename);
		std::stringstream stringstreamBuffer;
		std::string numberString;
		std::string stringBuffer;
		size_t pos;

		stringstreamBuffer << file.rdbuf();
		stringBuffer = stringstreamBuffer.str();

		while ((pos = stringBuffer.find('\n')) != std::string::npos)
		{
			data.push_back(std::stoi(stringBuffer.substr(0, pos)));
			stringBuffer.erase(0, pos + 1);
		}
		data.push_back(std::stoi(stringBuffer));
	}

	//std::stoi(buffer.str());

	switch (_part)
	{
	case 1:
		return Day1_1(data);
	case 2:
		return Day1_2(data);
	default:
		return -1;
	}
}

int Day1_1(std::vector<int> _dayData)
{
	int result = 0;
	int previous = _dayData[0];
	for (int current : _dayData)
	{
		if (current > previous)
		{
			result++;
		}
		previous = current;
	}
	return result;
}

int Day1_2(std::vector<int> _dayData)
{
	int result = 0;
	int previous = std::accumulate(_dayData.begin(), _dayData.begin() + 3, 0);
	for (int i = 0; i < _dayData.size(); i++)
	{
		int actual = i + 3 > _dayData.size() ? 0 : std::accumulate(_dayData.begin() + i, _dayData.begin() + i + 3, 0);
		if (actual > previous)
		{
			result++;
		}
		previous = actual;
	}
	return result;
}

File: AOC2021.cs/Day1.cs
using System;
using System.IO;
using System.Linq;

namespace AOC2021.cs
{
    internal class Day1
    {
        public static int Resolve(int _part, string _filename)
        {
            int[] data;
            try
            {
                data = File.ReadAllText(_filename)
                    .Split('\n')
                    .Select(s =>
                    {
                        int n;
                        return int.TryParse(s, out n) ? n : 0;
                    })
                    .ToArray();
            }
            catch (Exception)
            {

                return -1;
            }

            switch (_part)
            {
                case 1:
                    return Day1_1(data);
                case 2:
                    return Day1_2(data);
                default:
                    return -1;
            }
        }

        static int Day1_1(int[] _dayData)
        {
            int result = 0;
            int previous = _dayData[0];
            foreach (int current in _dayData)
            {
                if (current > previous)
                {
                    result++;
                }
                previous = current;
            }
            return result;
        }

        static int Day1_2(int[] _dayData)
        {
            int result = 0;
            int previous = _dayData.SubArray(0, 3).Sum();
            for (int i = 0; i < _dayData.Length; i++)
            {
                int actual = i + 3 > _dayData.Length ? 0 : _dayData.SubArray(i, 3).Sum();
                if (actual > previous)
                {
                    result++;
                }
                previous = actual;
            }
            return result;
        }
    }
}
